x-pack/plugins/fleet/server/services/package_policy.ts

```diff
@@ -620,29 +625,44 @@ class PackagePolicyService {
     return Promise.all(inputsPromises);
   }
 
+  public async runExternalCallbacks<A extends ExternalCallback[0]>(
+    externalCallbackType: A,
+    packagePolicy: NewPackagePolicy | DeletePackagePoliciesResponse,
+    context: RequestHandlerContext,
+    request: KibanaRequest
+  ): Promise<A extends 'postPackagePolicyDelete' ? void : NewPackagePolicy>;
   public async runExternalCallbacks(
     externalCallbackType: ExternalCallback[0],
-    newPackagePolicy: NewPackagePolicy,
+    packagePolicy: NewPackagePolicy | DeletePackagePoliciesResponse,
     context: RequestHandlerContext,
     request: KibanaRequest
-  ): Promise<NewPackagePolicy> {
-    let newData = newPackagePolicy;
-
-    const externalCallbacks = appContextService.getExternalCallbacks(externalCallbackType);
-    if (externalCallbacks && externalCallbacks.size > 0) {
-      let updatedNewData: NewPackagePolicy = newData;
-      for (const callback of externalCallbacks) {
-        const result = await callback(updatedNewData, context, request);
-        if (externalCallbackType === 'packagePolicyCreate') {
-          updatedNewData = NewPackagePolicySchema.validate(result);
-        } else if (externalCallbackType === 'packagePolicyUpdate') {
-          updatedNewData = UpdatePackagePolicySchema.validate(result);
+  ): Promise<NewPackagePolicy | void> {
+    if (externalCallbackType === 'postPackagePolicyDelete') {
+      const externalCallbacks = appContextService.getExternalCallbacks(externalCallbackType);
+      if (externalCallbacks && externalCallbacks.size > 0) {
+        for (const callback of externalCallbacks) {
+          await callback(packagePolicy as DeletePackagePoliciesResponse, context, request);
```

Comment:
is there a better way to type narrow this instead of forcing the type? If `packagePolicy` can be `NewPackagePolicy | DeletePackagePoliciesResponse` there is surely a way to guarantee it is one or the other.

Reply:
It should depend on the other variable (`externalCallbackType`). I was able to fix this for the return function type but not for this function param so I decided to use assertion. Do you know how we can consider the `packagePolicy` type depending on `externalCallbackType`?

Reply:
If you know that when `externalCallbackType` is A `packagePolicy` is X, you can do a validation for one unique `packagePolicy` attribute and that will narrow it and prevent this code to allow future developers mistakes returning data to this function. The else clause of that `packagePolicy` will never (in theory) execute, so you can use an assertNever so it gives a compilation error but never a runtime error.

Reply:
Done: https://github.com/elastic/kibana/pull/107507/commits/7ec935ce4054ab574e053cb4609e37ee2e9439f0